File: apps/web/src/lib/api/grammar.ts
import { apiClient } from "./client"

export interface GrammarTokenInput {
  w: string
  l: string
  pos: string
  x?: string    // language-specific POS tag (xpos)
  g?: string    // grammatical gender
  feats: string
  dep_head?: number
  dep_rel?: string
}

export interface TokenAnnotation {
  w: string
  annotation: string
}

export interface GrammarExplainResponse {
  token_annotations: TokenAnnotation[]
  prose_explanation: string
}

export async function explainGrammar(
  tokens: GrammarTokenInput[],
  languageCode: string,
  register?: string | null
): Promise<GrammarExplainResponse> {
  return apiClient<GrammarExplainResponse>("/grammar/explain", {
    method: "POST",
    body: JSON.stringify({ tokens, language_code: languageCode, register: register ?? null }),
  })
}

File: apps/backend/src/api/schemas/grammar.py
from pydantic import BaseModel


class TokenInput(BaseModel):
    w: str
    l: str
    pos: str
    x: str = ""   # language-specific POS tag (xpos)
    feats: str = ""
    dep_head: int = 0   # 1-based head index (0 = root)
    dep_rel: str = ""   # UD relation label, e.g. "nsubj", "obj"


class TokenAnnotation(BaseModel):
    w: str
    annotation: str


class GrammarExplainRequest(BaseModel):
    tokens: list[TokenInput]
    language_code: str
    register: str | None = None


class GrammarExplainResponse(BaseModel):
    token_annotations: list[TokenAnnotation]
    prose_explanation: str
